lang/clang-11-bootstrap/Portfile

```diff
@@ -0,0 +1,275 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:filetype=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           cmake 1.1
+PortGroup           muniversal 1.0
+
+# clang-11 is the last clang which can be build with the last version of cmake to not require c++11.
+name                clang-11-bootstrap
+
+version             11.1.0
+revision            0
+epoch               0
+
+platforms           darwin
+categories          lang
+maintainers         {@catap korins.ky:kirill} openmaintainer
+
+license             NCSA
+
+description         clang-11 with minimal MacPorts patches pegged at version ${version}.
+long_description    ${description} This port depends only on bootstrap dependencies.
+
+homepage            https://llvm.org/
+
+
+master_sites        https://github.com/llvm/llvm-project/releases/download/llvmorg-${version}
+distname            llvm-project-${version}.src
+
+checksums           rmd160  f566b4b75c8f30418f19069a9a84864ead766401 \
+                    sha256  74d2529159fd118c3eac6f90107b5611bccc6f647fdea104024183e8d5e25831 \
+                    size    84065492
+
+# no Clang / LLVM for PowerPC yet
+supported_archs     arm64 i386 x86_64
+
+depends_extract     port:xz-bootstrap
+depends_skip_archcheck-append \
+                    xz-bootstrap
+
+extract.suffix      .tar.xz
+extract.cmd         ${prefix}/libexec/xz-bootstrap/bin/xz
+
+depends_build-append \
+                    port:python27-bootstrap
+depends_skip_archcheck-append \
+                    python27-bootstrap
+
+# gcc10-bootstrap is used as compiler and as libc++
+# to avoid such dependency this port should also build libc++
+# and use 2-stage bootstrap
+depends_build-append \
```

Comment:
I really don't like the idea of always using

```
# OS X Yosemite default Xcode is new enough to compile directly
if {${os.major} < 14} {
    # gcc10-bootstrap is used as compiler and as libc++
    # to avoid such dependency this port should also build libc++
    # and use 2-stage bootstrap
    depends_build-append \
                    port:gcc10-bootstrap
}
```

This would speed things up on more recent versions of OS X/macOS

Reply:
Then also put the other libgcc block behind a similar block

Reply:
@Gcenx it makes things simpler and I really don't see any reason why someone is needed `clang-11-bootstrap` on something after Yosemite.

Reply:
Getting directly to modern cctools/ld64 this would be required to avoid the usual toolchain dance that currently happens within macports

Reply:
Thus, it also depends on python27-bootstrap on all systems, but it really needed such dependency on a few.

Reply:
> Getting directly to modern cctools/ld64 this would be required to avoid the usual toolchain dance that currently happens within macports

yes, but I prefer to keep things simpler here.

And if I may have one simpler way to build it everywhere, let use it.

Reply:
For an initial run I guess that makes sense, getting a stage2 build that’s not locked go libgcc would be preferable.

Reply:
@Gcenx I see this compiler as a tool to build `libcxx` where it is needed which allows to build `cmake` and `clang` and everything via usual toolset.

But it is the next step.

Reply:
`gcc10-bootstrap` > `clang-11-bootstrap` > `libcxx` > `clang-11-bootstrap` (stage2) > `cctools`/`lds64` > `clang`

I believe `clang-11-bootstrap` could be used to break the current clang dependency hell when attempting to get to a more modern macports-clang

Reply:
> You may build cctools and lds64 by clang-11-bootstrap.

Already been doing this since you'd first proposed this

A stage2 version of clang-11-bootstrap would ensure it no longer required gcc10-bootstrap, we'd first build lbcxx (on 10.6 and below) then rebuild clang while forcing it to use libcxx over libgcc. I think any c++ projects build using this will end up getting permanently linked to libgcc.

Reply:
it will…but we had decided before that having clang11-bootstrap linked to gcc10-bootstrap was **OK**. 

If you do a 2 stage build of clang11-bootstrap you need to be very careful which libcxx you are building against and exactly which libcxx you intend to install.

up to now, we install on 10.6 a specially modified libcxx from llvm 5.0 that works. There are reasons tonthink that installing anything newer that the libcc from llvm 7 might caise trouble (ABI change), but it might be OK.

needs testing.

Reply:
@Gcenx I feel that `stage2 clang` is usual `clang-X` port in macports. Am I wrong?

Reply:
@catap no Macports clang ports are not stage2 compilers.

Currently the only compilers that actually go through the bootstrapping are crossgcc (manually stages) and gcc10-bootstrap (using the bootstrap target)

Reply:
> @kencu the stage2 clang would be built against the system install libcxx or in 10.6 and belows cases this would be the custom libcxx 5.0
> 
> Being permanently linked to gcc10-bootstrap will be irritating considering things like cctools/ld64 end up being permanently linked to the LTO library from the compiler.
> 
> The idea is to use this compiler to break the current compiler toolchain dance that’s happening in macports.

Once again I feel I should not comment but -- 

clang11-bootstrap **cannot** depend on anything that it installs. Otherwise you are right back in the mess.

So IMHO leave clang11-bootstrap linked against gcc10-bootstrap. It will be much simpler in the end, I believe.

If it drives you crazy to have that space wasted, then you can copy the needed libgcc parts into clang11-bootstrap at the end of the build and change it to look for them there.

I read that clang is designed to look first in it's own lib trees for libgcc if it built against libgcc -- I haven't tried this, but if that is true, then this will be really easy.

Reply:
> clang11-bootstrap **cannot** depend on anything that it installs. Otherwise you are right back in the mess.

Exactly what I was talking about on macports-wine _if_ cctools/ld64 gets built with LTO it will be locked to clang-11-bootstrap. I guess that can be figured out later however as **libcxx** will be fine.

> So IMHO leave clang11-bootstrap linked against gcc10-bootstrap. It will be much simpler in the end, I believe.
> 
> If it drives you crazy to have that space wasted, then you can copy the needed libgcc parts into clang11-bootstrap at the end of the build and change it to look for them there.
> 
> I read that clang is designed to look first in it's own lib trees for libgcc if it built against libgcc -- I haven't tried this, but if that is true, then this will be really easy.

I'd also considered that but never gotten around to trying it, having libgcc copied into clang-11-bootstrap does sounds like the best possible compromise, at least then clang-11-bootstrap won't require gcc10-bootstrap just to function.

Reply:
> > clang11-bootstrap **cannot** depend on anything that it installs. Otherwise you are right back in the mess.
> 
> Exactly what I was talking about on macports-wine _if_ cctools/ld64 gets built with LTO it will be locked to clang-11-bootstrap. I guess that can be figured out later however as **libcxx** will be fine.

That can't happen like that then. 

cctools/ld64 can't be built against the llvm in clang11-bootstrap. 

clang11-bootstrap will have to build the llvm that is used by cctools/ld64 some other way. The system must function with clang11-bootstrap uninstalled. clang11-bootstrap is a build-time-only compiler, and can then be unistalled.

If a disposable cctools/ld64 is needed to bootstrap, then that will have to be used too. This is needed anyway, because otherwise you're hamstrung with bootstrapping ld64.

So still -- gcc10-bootstrap -> clang11-bootstrap -> builds everything else, starting with whatever is needed and going from there.

Reply:
running a double-build of clang11-bootstrap doesn't really solve any of this, AFAICT. You still wind up in loops depending on libc++. 

It is also plenty hard to do, passing all the fancy config stuff we need for building clang11 on 10.6 (and earlier) into the second bootstrap build is a big PITA (sure, I looked at this many times before)...

Reply:
Let’s ignore cctools/ld64 for the moment.

> running a double-build of clang11-bootstrap doesn't really solve any of this, AFAICT. You still wind up in loops depending on libc++

That’s only the case _if_ your expecting to do this in a single pass, separating the stages into its own subports that require the prior stage works around these kind of annoyances. Aka what’s already being done in the crossgcc group each compiler stage it’s it’s own port.